api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/LayerTest.kt

```diff
@@ -0,0 +1,142 @@
+        layer.build(tf, kGraph, inputShape)
+        val inputOp = getInputOp(tf, input)
+        val isTraining = tf.constant(true)
+        val numberOfLosses = tf.constant(1.0f)
+        val output = layer.forward(tf, inputOp, isTraining, numberOfLosses).asOutput()
+        return output
+    }
+
+    protected fun runLayerInEagerMode(
+        layer: Layer,
+        input: Array<*>,
+    ): Tensor<*> {
+        EagerSession.create().use {
+            val tf = Ops.create()
+            val kGraph = KGraph(Graph().toGraphDef())
+            val outputOp = getLayerOutputOp(tf, layer, input, kGraph)
+            return outputOp.tensor()
+        }
+    }
+
+    protected fun runLayerInGraphMode(
+        layer: Layer,
+        input: Array<*>,
+    ) : Tensor<*> {
+        Graph().use { graph ->
+            Session(graph).use { session ->
+                val tf = Ops.create(graph)
+                KGraph(graph.toGraphDef()).use { kGraph ->
+                    val outputOp = getLayerOutputOp(tf, layer, input, kGraph)
+                    /**
+                     * This check is necessary, otherwise it would throw the following error
+                     * if the layer has no trainable parameters:
+                     * ```
+                     * Must specify at least one target to fetch or execute.
+                     * ```
+                     * However, using `layer.weights.isNotEmpty()` for the condition would not work
+                     * because according to the current implementation of `extractWeights` the layer
+                     * should belong to an actual model and therefore it throws an error. Hence, we
+                     * are using `paramCount` instead to determine whether the layer has any trainable
+                     * weights.
+                     */
+                    if (layer.paramCount > 0)
+                        kGraph.initializeGraphVariables(session)
+                    val outputTensor = session.runner().fetch(outputOp).run().first()
+                    return outputTensor
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks the output of a layer given the input data is equal to the expected output.
+     *
+     * This takes care of building and running the layer instance ([layer]), in either of
+     * Eager or Graph mode execution ([runMode]) to verify the output of layer for the given
+     * input data ([input]), is equal to the expected output ([expectedOutput]).
+     *
+     * Note that this method could be used for a layer with any input/output dimensionality.
+     */
+    protected fun assertLayerOutputIsCorrect(
+        layer: Layer,
+        input: Array<*>,
+        expectedOutput: Array<*>,
+        runMode: RunMode = RunMode.EAGER,
+    ) {
+        val output = when (runMode) {
+            RunMode.EAGER -> runLayerInEagerMode(layer, input)
+            RunMode.GRAPH -> runLayerInGraphMode(layer, input)
+        }
```

Comment:
The `output` would be of type `Tensor<*>`, but I am not sure if I should call its `close()` method at the end of this function to release the allocated resources. This is what the `Tensor` documentation says:

> WARNING: Resources consumed by the Tensor object must be explicitly freed by invoking the close() method when the object is no longer needed. For example, using a try-with-resources block:
> ```
> try (Tensor t = Tensor.create(...)) {
>     doSomethingWith(t);
> }
> ```

And this is what the documentation of `close()` method says:

> WARNING:This must be invoked for all tensors that were not been produced by an eager operation or memory will be leaked.
> The Tensor object is no longer usable after close returns.

Reply:
better to do all closing of tensors manually, also the section `.use` could be used as an analogue of java try-with-resources

Reply:
also, you should close it for the graph session, of course, and test the closing for eager mode (maybe we could avoid closing for it), maybe you will got an error about the tensor was closed twice or something related incorrect tensor handling